File: src/test/kotlin/school/charset/app/domain/exercise/generator/Utf32ExerciseGeneratorTest.kt
package school.charset.app.domain.exercise.generator

import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import school.charset.app.domain.encoding.CodePoint
import school.charset.app.domain.encoding.Codec
import school.charset.app.domain.encoding.Encoding
import school.charset.app.domain.exercise.Step

class Utf32ExerciseGeneratorTest :
    FreeSpec({
        val codec = Codec()
        val utf32Generator = Utf32Generator(codec)
        val codePointGenerator = mockk<CodePointGenerator>()
        val byteArrayGenerator = mockk<ByteArrayGenerator>()

        val generator = Utf32ExerciseGenerator(
            codec = codec,
            encoding = Encoding.Utf32Be,
            codePointGenerator = codePointGenerator,
            byteArrayGenerator = byteArrayGenerator,
            utf32Generator = utf32Generator,
        )

        fun List<Step>.types(): List<String> = map { it::class.simpleName.orEmpty() }

        "the exercise never asks the learner to pick the endianness (it's given in the header)" {
            every { codePointGenerator.randomUtf32(Utf32Level.Bmp) } returns CodePoint(0xE9)
            every { codePointGenerator.randomUtf32(Utf32Level.Supplementary) } returns CodePoint(0x1F389)

            generator.generateEncode(1).steps.none { it is Step.Endianness } shouldBe true
            generator.generateEncode(2).steps.none { it is Step.Endianness } shouldBe true
        }

        "encode is binary (32-bit) then hex bytes - same shape for BMP and supplementary" {
            every { codePointGenerator.randomUtf32(Utf32Level.Bmp) } returns CodePoint(0xE9)

            val steps = generator.generateEncode(1).steps

            steps.types() shouldContainExactly listOf("Binary", "HexBytes")
            (steps[0] as Step.Binary).length shouldBe 32
            (steps[1] as Step.HexBytes).expected.shouldContainExactly(0x00, 0x00, 0x00, 0xE9)
        }

        "decode is binary (32-bit) then code point" {
            every { byteArrayGenerator.randomUtf32(Utf32Level.Bmp, Encoding.Utf32Be) } returns
                byteArrayOf(0x00, 0x00, 0x00, 0xE9.toByte())

            val steps = generator.generateDecode(1).steps

            steps.types() shouldContainExactly listOf("Binary", "CodePointEntry")
            (steps[0] as Step.Binary).length shouldBe 32
            (steps[1] as Step.CodePointEntry).expected shouldBe 0xE9
        }
    })

File: test/charset/exercise/generator/utf32_exercise_generator_test.exs
defmodule Charset.Exercise.Generator.Utf32ExerciseGeneratorTest do
  use ExUnit.Case, async: true

  alias Charset.Exercise.Generator.Utf32ExerciseGenerator
  alias Charset.Exercise.GenerationError
  alias Charset.Exercise.Step

  test "the exercise never asks the learner to pick the endianness (it's given in the header)" do
    refute 0xE9
           |> Utf32ExerciseGenerator.exercise_encode_steps(:utf32be)
           |> Enum.any?(&match?(%Step.Endianness{}, &1))

    refute <<0x00, 0x00, 0x00, 0xE9>>
           |> Utf32ExerciseGenerator.exercise_decode_steps(0xE9, :utf32be)
           |> Enum.any?(&match?(%Step.Endianness{}, &1))
  end

  test "encode keeps Binary + HexBytes" do
    assert [
             %Step.Binary{length: 32},
             %Step.HexBytes{expected: [0x00, 0x00, 0x00, 0xE9]}
           ] = Utf32ExerciseGenerator.exercise_encode_steps(0xE9, :utf32be)
  end

  test "decode keeps Binary + CodePointEntry" do
    assert [
             %Step.Binary{length: 32},
             %Step.CodePointEntry{expected: 0xE9}
           ] =
             Utf32ExerciseGenerator.exercise_decode_steps(
               <<0x00, 0x00, 0x00, 0xE9>>,
               0xE9,
               :utf32be
             )
  end

  describe "generate" do
    test "carries the requested encoding and level" do
      exercise = Utf32ExerciseGenerator.generate_encode!(:utf32le, 2)

      assert exercise.direction == :encode
      assert exercise.encoding == :utf32le
      assert exercise.level == 2
    end

    test "decode bytes roundtrip to the code point" do
      exercise = Utf32ExerciseGenerator.generate_decode!(:utf32le, 1)
      assert Charset.Encoding.Codec.decode!(exercise.bytes, :utf32le) == exercise.code_point
    end

    test "invalid level raises GenerationError" do
      error =
        assert_raise GenerationError, fn ->
          Utf32ExerciseGenerator.generate_decode!(:utf32le, 0)
        end

      assert Exception.message(error) ==
               "Cannot generate exercise for utf-32le level 0: level must be one of: 1, 2"
    end
  end
end
